### PR TITLE
Flow: timeslots related features with MongoDB backend

### PR DESCRIPTION
**Pull request by vruello**
This PR makes possible to reduce timeslots precisions in the flow purpose **with MongoDB backend**. It also adds new temporal filters to get flows that have been seen after or before a given date (based on timeslots).

The timeline of the Web UI has been reworked. The "max_timeslots" option has been removed to enforce consistency in timeline display: it is only based on timeslots precision and nothing else. In result, for very long periods or very small timeslots durations, the rendering of the timeline can be slow. In that case, you should reduce timeslots precision. 

When using `bro2db`, timeslots precision is still defined by `config.FLOW_TIME_PRECISION`, which represents the duration of each timeslot in seconds and which defaults to 3600 (1 hour).

By the same time, this PR improves the consistency of the code by putting prototypes of backend independent methods in `db/__init__.py`.

### CODE REVIEW

**Comment by p-l-** [comment]
Hi,
Looks like you need to rebase this one against the current master.
Thanks!

File: ivre/db/mongo.py [hunk 1]
```diff
@@ -5040,6 +5039,7 @@ def topvalues(self, flt, fields, collect_fields=None, sum_fields=None,
 
             res_count = ext_entry.pop('_count')
             # Format collected results in a set of tuples to avoid duplicates
+            res_collected = set()
```
Comment (vruello) [inline ivre/db/mongo.py]: This is not directly related to this PR, but it fixes an issue which happens when using `topvalues` without collected values: `res_collected` must be initialized even if there are no collected values.

Comment (p-l-) [reply ivre/db/mongo.py]: You could move that to an `else` after the following `if` statement?

File: ivre/web/app.py [hunk 2]
```diff
@@ -582,6 +583,17 @@ def get_flow():
     count = query.get("count", False)
     orderby = query.get("orderby", None)
     timeline = query.get("timeline", False)
+    try:
+        before = datetime.datetime.strptime(query.get("before", None),
```
Comment (p-l-) [inline ivre/web/app.py]: `, None` is not useful here. Also, maybe it could be good to have a debug / warning message when before is provided incorrectly.
```Python
    try:
        before = datetime.datetime.strptime(query["before"], "%Y-%m-%d %H:%M")
    except KeyError:
        before = None
    except ValueError:
        # XXX warning
        before = None
```

File: ivre/db/__init__.py [hunk 3]
```diff
@@ -2448,11 +2448,9 @@ def date_round(cls, date):
 
     @classmethod
     def from_filters(cls, filters, limit=None, skip=0, orderby="", mode=None,
-                     timeline=False):
+                     timeline=False, after=None, before=None, precision=None):
```
[the remark this answers is not in the document]
Comment (vruello) [reply ivre/db/__init__.py]: Even thought these parameters are currently unused in Neo4j backend, they are used in the MongoDB backend. Therefore, I don't think they should be marked as unused in the backend independent interface.

In addition, a docstring states that they are unused in neo4j.py:1237.

File: ivre/db/__init__.py [hunk 4]
```diff
@@ -2467,16 +2465,135 @@ def _get_timeslots(cls, start_time, end_time):
...
+        first_timeslot = cls._get_timeslot(
+            start_time, config.FLOW_TIME_PRECISION, config.FLOW_TIME_BASE
+        )
+        time = first_timeslot['start']
+        last_timeslot = cls._get_timeslot(
+            end_time, config.FLOW_TIME_PRECISION, config.FLOW_TIME_BASE
+        )
+        end_time = last_timeslot['start']
+        while time <= end_time:
             d = OrderedDict()
```
[the remark this answers is not in the document]
Comment (vruello) [reply ivre/db/__init__.py]: In Python, each timeslot is a dictionary like `{'start': <datetime>, 'duration': <int>}`. Dictionary type does not exist in MongoDB. Thus, before being inserted in MongoDB, these dictionaries are transformed in `Object` by `pymongo`. 

Timeslots are stored as an `Array` in MongoDB. To insert/update timeslots, I use the MongoDB operator `$addToSet` which ensures the unicity of elements. `Object` elements are sensible to the attributes order, meaning that `{a: 1, b: 2}` is different than `{b: 2, a: 1}`. 

Because in Python the order of fields in a dictionary is undefined, I must use an OrderedDict so that I don't get duplicated timeslots.

PS : It is to be noted that Python 2.6 does not support `OrderedDict`. The fallback is to use a classic `dict`. According to Travis builds (actually the absence of random fails for Python 2.6), it seems that Python 2.6 dicts are ordered by default.

File: ivre/db/__init__.py [hunk 5]
```diff
@@ -2467,16 +2465,135 @@ def _get_timeslots(cls, start_time, end_time):
...
             time += timedelta(seconds=config.FLOW_TIME_PRECISION)
         return times
 
+    @staticmethod
+    def _get_timeslot(time, precision, base):
+        ts = utils.datetime2timestamp(time)
+        ts += utils.tz_offset(ts)
+        new_ts = ts - (((ts % precision) - base) % precision)
+        new_ts -= utils.tz_offset(new_ts)
+        d = OrderedDict()
```
[the remark this answers is not in the document]
Comment (vruello) [reply ivre/db/__init__.py]: You're right :)

File: ivre/db/__init__.py [hunk 6]
```diff
@@ -2467,16 +2465,135 @@ def _get_timeslots(cls, start_time, end_time):
...
+        d["start"] = datetime.fromtimestamp(new_ts)
+        d["duration"] = precision
+        return d
+
+    def reduce_precision(self, new_duration, flt=None,
```
[the remark this answers is not in the document]
Comment (vruello) [reply ivre/db/__init__.py]: You're right. 

By the way, I've had some doubts with naming this "precision" everywhere, because it refers to the periods duration.

File: ivre/db/__init__.py [hunk 7]
```diff
@@ -2467,16 +2465,135 @@ def _get_timeslots(cls, start_time, end_time):
...
+            - the given filter <flt> if specified
+            - that have been seen before <before> if specified
+            - that have been seen after <after> if specified
+            - timeslots changed must currently have <precision> precision if
+                specified
+        <base> represents the timestamp of the base point.
+        If <precision> is specified:
+            <new_duration> must be a multiple of <precision>
+            <new_duration> must be greater than <precision>
+            <base> must be a multiple of current precision
```
[the remark this answers is not in the document]
Comment (vruello) [reply ivre/db/__init__.py]: Actually `base` should not have been kept here because it has been removed from the MongoDB implementation. Instead, I get the base from the config (`FLOW_TIME_BASE`). Otherwise, I would have had to support multiple bases, leading to store in each timeslot its base. It seems like an unnecessary overhead.

File: ivre/db/neo4j.py [hunk 8]
```diff
@@ -1268,14 +1285,17 @@ def count(self, query):
...
+        Returns a generator within each element is a dict
         {
             flows: [("proto/dport", count), ...]
             time_in_day: time
         }
-        WARNING/FIXME: this mutates the query
+        WARNING/FIXME: this mutates the query.
+        Note: precision, after, before are IGNORED in neo4j backend.
         """
+        query = flt
```
[the remark this answers is not in the document]
Comment (vruello) [reply ivre/db/neo4j.py]: In MongoDB backend, `flt` is not compulsory but it is in Neo4j. To fix this, I've made this argument compulsory everywhere.

File: ivre/db/__init__.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (p-l-) [reply ivre/db/__init__.py]: Perfect. Could you add this explanation as a comment in the code?

File: ivre/db/__init__.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (p-l-) [reply ivre/db/__init__.py]: OK.
